Bugsnag/BugsnagErrorReportSink.m

```diff
@@ -0,0 +1,180 @@
+@end
+
+@interface BugsnagErrorReportSink ()
+@property NSMutableSet<NSString *> *activeRequests;
+@end
+
+@implementation BugsnagErrorReportSink
+
+- (instancetype)initWithApiClient:(BugsnagErrorReportApiClient *)apiClient {
+    if (self = [super init]) {
+        self.apiClient = apiClient;
+        self.activeRequests = [NSMutableSet new];
+    }
+    return self;
+}
+
+/**
+ * Returns a list of filenames which have no ongoing request and need to be delivered to the error reporting API.
+ *
+ * To prevent duplicate reports filenames are recorded in an internal collection when they are
+ * in the process of being sent in a request. This method adds any unqueued files to this internal collection.
+ * Once the request has complete the filename should be removed via a separate method.
+ */
+- (NSArray<NSString *> *)prepareNewRequests:(NSArray<NSString *> *)allRequests {
+    NSMutableArray *newRequests = [allRequests mutableCopy];
+    @synchronized (self.activeRequests) {
+        NSArray *enqueuedRequests = [self.activeRequests allObjects];
+        [newRequests removeObjectsInArray:enqueuedRequests];
+        [self.activeRequests addObjectsFromArray:newRequests];
+    }
+    return newRequests;
+}
+
+- (void)finishActiveRequest:(NSString *)requestId
+                    success:(BOOL)success
+                      error:(NSError *)error
+                      block:(BSGOnErrorSentBlock)block {
+    block(requestId, success, error);
+    @synchronized (self.activeRequests) {
+        [self.activeRequests removeObject:requestId];
+    }
+}
+
+- (void)sendStoredReports:(NSDictionary <NSString *, NSDictionary *> *)ksCrashReports
+                withBlock:(BSGOnErrorSentBlock)block {
+                    
+    // 1. check whether filenames are in the dictionary, if so add them
+    // 2. If not, add them and enqueue a request
+    // 3. If so, ignore them and perform no action
+    // 4. When a request has completed and deleted the file, remove the files from the dictionary
+    NSArray<NSString *> *keys = [self prepareNewRequests:[ksCrashReports allKeys]];
+    NSMutableDictionary<NSString *, BugsnagEvent *>* storedEvents = [NSMutableDictionary new];
+    BugsnagConfiguration *configuration = [Bugsnag configuration];
+
+    for (NSString *fileKey in keys) {
+        NSDictionary *report = ksCrashReports[fileKey];
+        BugsnagEvent *event = [[BugsnagEvent alloc] initWithKSReport:report];
+        event.codeBundleId = [Bugsnag client].codeBundleId;
+
+        if ([event shouldBeSent] && [self runOnSendBlocks:configuration event:event]) {
+            storedEvents[fileKey] = event;
+        } else { // delete the report as the user has discarded it
+            [self finishActiveRequest:fileKey success:true error:nil block:block];
+        }
+    }
+    [self deliverStoredEvents:storedEvents configuration:configuration block:block];
+}
+
+- (void)deliverStoredEvents:(NSMutableDictionary<NSString *, BugsnagEvent *> *)storedEvents
```

Comment:
We discuss IRL whether this should be done in separate phases - can you comment the function with some rationale?